backend/operations_scanner.go

```diff
@@ -0,0 +1,346 @@
+		if code == "" {
+			code = arm.CloudErrorCodeInternalServerError
+		}
+		message := clusterStatus.ProvisionErrorMessage()
+		if message == "" {
+			message = clusterStatus.Description()
+		}
+		opError = &arm.CloudErrorBody{Code: code, Message: message}
+	case cmv1.ClusterStateInstalling:
+		opStatus = arm.ProvisioningStateProvisioning
+	case cmv1.ClusterStateReady:
+		opStatus = arm.ProvisioningStateSucceeded
+	case cmv1.ClusterStateUninstalling:
+		opStatus = arm.ProvisioningStateDeleting
+	case cmv1.ClusterStatePending, cmv1.ClusterStateValidating:
+		// These are valid cluster states for ARO-HCP but there are
+		// no unique ProvisioningState values for them. They should
+		// only occur when ProvisioningState is Accepted.
+		if opStatus != arm.ProvisioningStateAccepted {
+			logger.Warn(fmt.Sprintf("Got ClusterState '%s' while ProvisioningState was '%s' instead of '%s'", clusterStatus.State(), opStatus, arm.ProvisioningStateAccepted))
+		}
+		return
+	default:
+		logger.Warn(fmt.Sprintf("Unhandled ClusterState '%s' for operation '%s'", clusterStatus.State(), doc.ID))
+		return
+	}
+
+	err = s.withSubscriptionLock(ctx, logger, doc.OperationID.SubscriptionID, func(ctx context.Context) error {
+		return s.updateOperationStatus(ctx, logger, doc, opStatus, opError)
+	})
+
+	return
+}
+
+func (s *OperationsScanner) pollNodePoolOperation(ctx context.Context, logger *slog.Logger, doc *database.OperationDocument) (requeue bool, err error) {
+	// FIXME Implement when new OCM API is available.
+	requeue = true
+	return
+}
+
+func (s *OperationsScanner) withSubscriptionLock(ctx context.Context, logger *slog.Logger, subscriptionID string, fn func(ctx context.Context) error) error {
+	timeout := s.lockClient.GetDefaultTimeToLive()
+	lock, err := s.lockClient.AcquireLock(ctx, subscriptionID, &timeout)
+	if err != nil {
+		return fmt.Errorf("Failed to acquire lock for subscription '%s': %w", subscriptionID, err)
+	}
+
+	lockedCtx, stop := s.lockClient.HoldLock(ctx, lock)
+	err = fn(lockedCtx)
+	lock = stop()
+
+	if lock != nil {
+		nonFatalErr := s.lockClient.ReleaseLock(ctx, lock)
+		if nonFatalErr != nil {
+			// Failure here is non-fatal but still log the error.
+			// The lock's TTL ensures it will be released eventually.
+			logger.Error(fmt.Sprintf("Failed to release lock for subscription '%s': %v", subscriptionID, nonFatalErr))
+		}
+	}
+
+	return err
+}
+
+func (s *OperationsScanner) deleteOperationCompleted(ctx context.Context, logger *slog.Logger, doc *database.OperationDocument) error {
+	err := s.dbClient.DeleteResourceDoc(ctx, doc.ExternalID)
+	if err != nil {
+		return err
+	}
+
+	// Save a final "succeeded" operation status until TTL expires.
+	const opStatus arm.ProvisioningState = arm.ProvisioningStateSucceeded
+	updated, err := s.dbClient.UpdateOperationDoc(ctx, doc.ID, func(updateDoc *database.OperationDocument) bool {
+		if opStatus != updateDoc.Status {
+			updateDoc.LastTransitionTime = time.Now()
```

Comment:
You likely want to store `now func() time.Time` as a member on `OperationsScanner` to allow any of this to be unit-tested while injecting a fake clock.

Reply:
(global comment for the whole PR)